File: 841-keys-and-rooms/cpp/main.cpp
class Solution {
 public:
  bool canVisitAllRooms(vector<vector<int>>& rooms) {
    unordered_set<int> unvisited, visited;
    unvisited.insert(0);

    while (!unvisited.empty()) {
      int visiting = *unvisited.begin();
      unvisited.erase(unvisited.begin());
      visited.insert(visiting);
      for (int next : rooms[visiting]) {
        if (visited.find(next) == visited.end()) unvisited.insert(next);
      }
    }

    return visited.size() == rooms.size();
  }
};

File: 841-keys-and-rooms/go/main.go
package main

func canVisitAllRooms(rooms [][]int) bool {
	unvisited := make(map[int]bool)
	visited := make(map[int]bool)

	unvisited[0] = true

	for len(unvisited) != 0 {
		visiting := -1
		for k, v := range unvisited {
			if v {
				visiting = k
				break
			}
		}

		delete(unvisited, visiting)
		visited[visiting] = true
		for _, i := range rooms[visiting] {
			if !visited[i] {
				unvisited[i] = true
			}
		}
	}

	return len(visited) == len(rooms)
}
